openforcefield/topology/topology.py

```diff
@@ -0,0 +1,2190 @@
+        # Reverse the key if the first element is bigger than the last.
+        if key[0] > key[-1]:
+            key = tuple(reversed(key))
+        return key
+
+class ImproperDict(_TransformedDict):
+    """Symmetrize improper torsions"""
+    def __keytransform__(self,key):
+        """Reorder tuple in numerical order except for element[1] which is the central atom; it retains its position."""
+        # Ensure key is a tuple
+        key = tuple(key)
+        # Retrieve connected atoms
+        connectedatoms = [key[0], key[2], key[3]]
+        # Sort connected atoms
+        connectedatoms.sort()
+        # Re-store connected atoms
+        key = tuple( [connectedatoms[0], key[1], connectedatoms[1], connectedatoms[2]])
+        return(key)
+
+#=============================================================================================
+# TOPOLOGY OBJECTS
+#=============================================================================================
+
+#=============================================================================================
+# Particle
+#=============================================================================================
+
+class Particle(object):
+    """
+    Base class for all particles in a molecule.
+
+    A particle object could be an ``Atom`` or a ``VirtualSite``.
+
+    """
+    def __init__(self, name):
+        """
+        Create a particle.
+        """
+        self._name = name # the particle name
+        self._topology = None # the Topology object this Particle belongs to
+
+    @property
+    def topology(self):
+        """
+        The Topology object that owns this particle, or None.
+        """
+        return self._topology
+
+    @property
+    def name(self):
+        """
+        An arbitrary label assigned to the particle.
+
+        """
+        return self._name
+
+    @property
+    def particle_index(self):
+        """
+        Index of this particle within the ``Topology`` or corresponding OpenMM ``System`` object.
+
+        .. todo::
+
+           Should ``atom.particle_index`` just be called ``index``, or does that risk confusion within
+           the index within ``topology.atoms``, which will differ if the system has virtual sites?
+
+        """
+        if self._topology is None:
+            raise Exception('This particle does not belong to a Topology')
+        # Return index of this particle within the Topology
+        # TODO: This will be slow; can we cache this and update it only when needed?
+        #       Deleting atoms/molecules in the Topology would have to invalidate the cached index.
+        return self._topology.particles.index(self)
+
+    def __repr__(self):
+        pass
+
+    def __str__(self):
+        pass
+
+#=============================================================================================
+# Atom
+#=============================================================================================
+
+class Atom_unused(Particle):
```

Comment:
Should this be cleaned out?

Reply:
Or at least prepended with `_` to not make it part of the publicly exported API?